### PR TITLE
DEM-870 Search field component cypress tests and refactor

### PR DESCRIPTION
**Pull request by teodossidossev**

### CODE REVIEW

File: cypress/steps/search-field-steps.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
+export default class SearchFieldSteps {
+  static visit(useCustomTemplate: boolean = false) {
```
Comment (svilenvelikov) [inline cypress/steps/search-field-steps.ts]: I'd prefer to the tests and steps to be conditions free. Here you can just have two separate methods like `visit` and `visitWithTemplate` or something

Comment (teodossidossev) [reply cypress/steps/search-field-steps.ts]: fixed

File: cypress/test/search-field.spec.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    SearchFieldSteps.getStateSelection(0).should('be.visible').should('contain', 'Dehydroepiandrosterone');
+
+    SearchFieldSteps.getStateSelection(1).should('not.be.visible');
+  });
+
+  it.only('Should use custom template if passed', () => {
+    SearchFieldSteps.visitWithCustomTemplate();
+    SearchFieldSteps.typeSearchPhrase('D');
+
+    SearchFieldSteps.getCustomAutocompleteLabel(0).should('contain', 'Dehydroepiandrosterone');
```
Comment (yordanalexandrov) [inline cypress/test/search-field.spec.ts:63]: How is that different than when not using a template?

File: projects/onto-search/src/lib/onto-search-field/search/search.component.html [hunk 3]
```diff
@@ -19,16 +20,17 @@
...
           (matChipInputTokenEnd)="add($event)"
-          (input)="onInputChange($event.target.value)">
+          (input)="onInputChange($event.target.value)"
+          appCypressData="search-field-input">
       </mat-chip-list>
       <mat-autocomplete #auto="matAutocomplete" (optionSelected)="selected($event)">
-        <mat-option class="state-option" *ngFor="let state of filteredStates | async" [value]="state">
-          <ng-container *ngTemplateOutlet="currentTemplate; context: {$implicit: state}">
+        <mat-option class="state-option" *ngFor="let state of filteredStates | async; let index = index" [value]="state">
+          <ng-container *ngTemplateOutlet="currentTemplate; context: {$implicit: {state: state, index: index}}">
```
Comment (yordanalexandrov) [inline projects/onto-search/src/lib/onto-search-field/search/search.component.html:28]: You don't need to have everything as implicit context.
`context: {$implicit: {state}, index}` is also valid and you will have `let-item; let-index`

File: src/app/components/search-field/search-field.component.html [hunk 4]
```diff
@@ -0,0 +1,13 @@
+<onto-search-field
+  [autocompleteData]="states"
+  [customTemplate]="useCustomTemplate() ? customTemplate : null"
+  [selectedList]="getSelectedList()"
+  [enableInnerAutocompleteFiltration]="enableInnerAutocompleteFiltration()"
+  (onSearch)="onSearchEvent()"
+  (onKeyPress)="onKeyPressEvent()"></onto-search-field>
+
+<ng-template #customTemplate let-item>
```
Comment (yordanalexandrov) [inline src/app/components/search-field/search-field.component.html:21]: This looks like the default template.

File: src/app/components/search-field/search-field.component.html [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<onto-search-field
+  [autocompleteData]="states"
+  [customTemplate]="useCustomTemplate() ? customTemplate : null"
+  [selectedList]="getSelectedList()"
```
Comment (yordanalexandrov) [inline src/app/components/search-field/search-field.component.html]: When you pass a function to an Input it executes on every change detection.
What is the benefit here of passing functions? Do you change the values of the inputs dynamicaly?

File: src/app/components/search-field/search-field.component.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+  public useCustomTemplate(): boolean {
+    return this.customTemplate;
+  }
+
+  public getSelectedList(): [] {
+    return [];
+  }
+
+  public onSearchEvent(): void {
+
```
Comment (yordanalexandrov) [inline src/app/components/search-field/search-field.component.ts:138]: Should test if the event fires.

File: projects/onto-search/src/lib/onto-search-field/onto-search-field.component.ts [hunk 7]
```diff
@@ -2,24 +2,50 @@ import {Component, EventEmitter, Input, OnInit, Output, TemplateRef} from '@angu
...
   private autocompleteData: Observable<SearchModel[]>;
+
+  /**
+   * Configuration used to enable or disable inner autocomplete filtration. False by default if not passed.
+   */
+  @Input()
+  enableInnerAutocompleteFiltration: boolean;
+
+  /**
+   * On search event emitter
```
Comment (yordanalexandrov) [inline projects/onto-search/src/lib/onto-search-field/onto-search-field.component.ts]: By it's name I get that it is an on search event emitter, but when does it fire and what does it emit?

File: cypress/test/search-field.spec.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
+import SearchFieldSteps from "../../cypress/steps/search-field-steps";
+
+describe('Onto Search Results Component', () => {
```
Comment (svilenvelikov) [inline cypress/test/search-field.spec.ts]: Should be `Onto Search Field Component`

File: projects/onto-search/src/lib/onto-search-field/onto-search-field.component.ts [hunk 9]
```diff
@@ -2,24 +2,50 @@ import {Component, EventEmitter, Input, OnInit, Output, TemplateRef} from '@angu
...
   customTemplate?: TemplateRef<any>;
+
+  /**
+   * Preselected states list.
+   */
   @Input()
   private selectedList: any[];
+
+  /**
+   * Autocomplete response
```
Comment (yordanalexandrov) [inline projects/onto-search/src/lib/onto-search-field/onto-search-field.component.ts]: I don't like the comment. It is an input that is a response?
Something like `Observable of data that is provides autocomplete options` makes more sense to me.

File: cypress/test/search-field.spec.ts [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    SearchFieldSteps.getAutocompleteLabel(0).click();
+    SearchFieldSteps.completeSearchPhrase('free text');
+
+    SearchFieldSteps.getStateSelection(0).should('be.visible')
+      .should('contain', 'Dehydroepiandrosterone')
+      .should('have.css', 'background-color', 'rgb(173, 216, 230)');
+
+    SearchFieldSteps.getStateSelection(1).should('be.visible')
+      .should('contain', 'free text')
+      .should('have.css', 'background-color', 'rgb(224, 224, 224)');
```
Comment (svilenvelikov) [inline cypress/test/search-field.spec.ts]: Can't we have a css class for finding the free text selection and the regular once?

File: projects/onto-search/src/lib/onto-search-field/models/search-configuration.ts [hunk 11]
```diff
@@ -1,5 +1,8 @@
 import {COMMA, ENTER} from '@angular/cdk/keycodes';
 
+/**
+ * Configuration of the search input and mat-chip integration.
+ */
```
Comment (yordanalexandrov) [inline projects/onto-search/src/lib/onto-search-field/models/search-configuration.ts]: If SearchConfiguration is exposed, what does visible and selectable configure?

File: cypress/test/search-field.spec.ts [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+  });
+
+  it('Should delete selected', () => {
+    SearchFieldSteps.visit();
+    SearchFieldSteps.typeSearchPhrase('D');
+    SearchFieldSteps.getAutocompleteLabel(0).click();
```
Comment (svilenvelikov) [inline cypress/test/search-field.spec.ts:35]: maybe having a method `selectOption(index)` is a good idea

File: cypress/test/search-field.spec.ts [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+    SearchFieldSteps.getStateSelection(1).should('not.be.visible');
+  });
+
+  it('Should not duplicate selected', () => {
+    SearchFieldSteps.visit();
+    SearchFieldSteps.typeSearchPhrase('D');
+    SearchFieldSteps.getAutocompleteLabel(0).should('contain', 'Dehydroepiandrosterone').click();
+
+    SearchFieldSteps.getStateSelection(0).should('be.visible').should('contain', 'Dehydroepiandrosterone');
```
Comment (svilenvelikov) [inline cypress/test/search-field.spec.ts]: BTW there is an `and` alias of the `should` command which makes the things more readable

File: cypress/test/search-field.spec.ts [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+  it.only('Should use custom template if passed', () => {
+    SearchFieldSteps.visitWithCustomTemplate();
+    SearchFieldSteps.typeSearchPhrase('D');
+
+    SearchFieldSteps.getCustomAutocompleteLabel(0).should('contain', 'Dehydroepiandrosterone');
+    SearchFieldSteps.getCustomAutocompleteType(0).should('contain', 'small molecule');
+    SearchFieldSteps.getCustomAutocompleteLabels(0).should('contain', 'Prasterone,5-Dehydroepiandrosterone,DHEA,5-DHEA,3-beta-Hydroxy-5-androsten-17-one,DHA');
+  });
+
+
```
Comment (yordanalexandrov) [inline cypress/test/search-field.spec.ts:74]: Internal filtering is not tested

File: cypress/test/search-field.spec.ts (hunk 10 above)
Comment (svilenvelikov) [reply cypress/test/search-field.spec.ts]: I just saw that you you actually have such classes. Why don't you just check for the class and not for these colors here? It's more fragile as it is now.

File: projects/onto-search/src/lib/onto-search-field/search/search.component.scss [hunk 15]
```diff
@@ -1,4 +1,4 @@
-.search-component {
+.search-field-component {
```
Comment (svilenvelikov) [inline projects/onto-search/src/lib/onto-search-field/search/search.component.scss:1]: The button below can be targetted by its class and by tag

File: projects/onto-search/src/lib/onto-search-field/models/search-model.ts [hunk 16]
```diff
@@ -1,3 +1,7 @@
+/**
+ * Interface defining the minimal search response parameters.
+ * Search response  must implement this interface in order response data to be processed.
+ */
```
Comment (svilenvelikov) [inline projects/onto-search/src/lib/onto-search-field/models/search-model.ts]: We did a mistake with the renaming of the component. It's kind of inconsistent now. You have a SearchFieldComponent but SearchConfiguration and SearchModel. The renaming is not applied everywhere and leads to confusion.

File: projects/onto-search/src/lib/onto-search-field/onto-search-field.component.ts [hunk 17]
```diff
@@ -2,24 +2,50 @@ import {Component, EventEmitter, Input, OnInit, Output, TemplateRef} from '@angu
...
+   */
   @Input()
   private autocompleteData: Observable<SearchModel[]>;
+
+  /**
+   * Configuration used to enable or disable inner autocomplete filtration. False by default if not passed.
```
Comment (svilenvelikov) [inline projects/onto-search/src/lib/onto-search-field/onto-search-field.component.ts]: Maybe it's by me, but I just don't get it what's the purpose of this flag. And the comment doesn't say much more than the obvious. I'm just not too familiar with these material ui components.

File: cypress/test/search-field.spec.ts (hunk 2 above)
Comment (teodossidossev) [reply cypress/test/search-field.spec.ts:63]: different templates have different selectors

File: projects/onto-search/src/lib/onto-search-field/search/search.component.html (hunk 3 above)
Comment (teodossidossev) [reply projects/onto-search/src/lib/onto-search-field/search/search.component.html:28]: its a bit more readable to me

File: src/app/components/search-field/search-field.component.html (hunk 4 above)
Comment (teodossidossev) [reply src/app/components/search-field/search-field.component.html:21]: with different selectors only - IMO it's enough for testing

File: projects/onto-search/src/lib/onto-search-field/search/search.component.html (hunk 3 above)
Comment (yordanalexandrov) [reply projects/onto-search/src/lib/onto-search-field/search/search.component.html:28]: Most people implementing a custom template won't care about the index of option.

Comment (yordanalexandrov) [reply projects/onto-search/src/lib/onto-search-field/search/search.component.html:28]: Either way, template should be documented.

File: cypress/test/search-field.spec.ts [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+    SearchFieldSteps.typeSearchPhrase('D');
+    SearchFieldSteps.getAutocompleteLabel(0).should('contain', 'Dehydroepiandrosterone').click();
+
+    SearchFieldSteps.getStateSelection(0).should('be.visible').and('contain', 'Dehydroepiandrosterone');
+
+    SearchFieldSteps.typeSearchPhrase('D');
+    SearchFieldSteps.getAutocompleteLabel(0).should('contain', 'Dehydroepiandrosterone').click();
+    SearchFieldSteps.getStateSelection(0).should('be.visible').and('contain', 'Dehydroepiandrosterone');
+
+    SearchFieldSteps.getStateSelection(1).should('not.be.visible');
```
Comment (svilenvelikov) [inline cypress/test/search-field.spec.ts]: uhmm this is kind of confuing. Why don't you check for selections size instead?

Comment (teodossidossev) [reply cypress/test/search-field.spec.ts]: done

**Review by svilenvelikov — APPROVED** [review]

**Review by yordanalexandrov — APPROVED** [review]

**Pull request merged**